include/boost/numeric/ublas/tensor/detail/extents_functions.hpp

```diff
@@ -0,0 +1,261 @@
+    return false;
+  }
+
+  auto greater_one = [](auto const &a) {
+    return a > typename ExtentsType::value_type(1);
+  };
+
+  if constexpr (!detail::is_static_extents<ExtentsType>::value) {
+    return std::any_of(e.begin() + 2, e.end(), greater_one);
+  } else {
+    auto arr = e.to_array();
+    return std::any_of(arr.begin() + 2, arr.end(), greater_one);
+  }
+}
+
+/** @brief Eliminates singleton dimensions when size > 2
+ *
+ * squeeze {  1,1} -> {  1,1}
+ * squeeze {  2,1} -> {  2,1}
+ * squeeze {  1,2} -> {  1,2}
+ *
+ * squeeze {1,2,3} -> {  2,3}
+ * squeeze {2,1,3} -> {  2,3}
+ * squeeze {1,3,1} -> {  1,3}
+ *
+ * @returns basic_extents<int_type> with squeezed extents
+ */
+template <class ExtentsType, typename std::enable_if<detail::is_extents<ExtentsType>::value, int>::type = 0>
+auto squeeze(ExtentsType const &e) {
+  
+  if (e.size() <= 2) {
+    if constexpr (detail::is_static_extents<ExtentsType>::value) {
+      return e.to_dynamic_extents();
+    } else {
+      return e;
+    }
+  }
+
+  auto new_extent = typename ExtentsType::base_type();
+  auto old_extent = e.base();
+
+  auto insert_iter =
+      std::back_insert_iterator<typename ExtentsType::base_type>(new_extent);
+
+  auto c = std::count_if(old_extent.begin(), old_extent.end(), [](auto &n) {
+    return n == typename ExtentsType::value_type(1);
+  });
+
+  auto num = old_extent.size() - c;
+
+  if (old_extent[0] == typename ExtentsType::value_type(1) &&
+      old_extent[1] != typename ExtentsType::value_type(1) && num == 1u) {
+    new_extent.push_back(typename ExtentsType::value_type(1));
+    new_extent.push_back(old_extent[1]);
+  } else {
+    std::remove_copy(old_extent.begin(), old_extent.end(), insert_iter,
+                     typename ExtentsType::value_type(1));
+  }
+
+  while (new_extent.size() < 2)
+    new_extent.push_back(typename ExtentsType::value_type(1));
+
+  return basic_extents<typename ExtentsType::value_type>{new_extent};
+}
+
+/** @brief Returns the number of elements a tensor holds with this */
+template <class ExtentsType, typename std::enable_if<detail::is_extents<ExtentsType>::value, int>::type = 0>
+BOOST_UBLAS_TENSOR_INLINE constexpr auto product(ExtentsType const &e) {
+
+  if (e.empty()) {
+    return typename ExtentsType::value_type(0);
+  }
+
+  if constexpr( detail::is_static<ExtentsType>::value){
+    if constexpr (ExtentsType::Rank == 0) {
+      return typename ExtentsType::value_type(0);
+    } else {
+      return detail::product_helper<ExtentsType>::value;
+    }
+  }else {
+    if constexpr (!detail::is_static_extents<ExtentsType>::value) {
+      return typename ExtentsType::value_type(
+          std::accumulate(e.begin(), e.end(), 1ul, std::multiplies<>()));
+    } else {
+      auto p = typename ExtentsType::value_type(1);
+      for (auto i = 0; i < e.size(); i++) {
+        p *= e.at(i);
+      }
+      return p;
+    }
+  }
+}
+
+
+
+} // namespace boost::numeric::ublas
+#endif
```

Comment:
nit-pick: missing newline at eof.

Reply:
I will fix it.

Reply:
Resolving conversation as file is outdated.